parity-clib/parity.h

```diff
@@ -86,21 +88,52 @@ int parity_start(const ParityParams* params, void** out);
 ///					must not call this function.
 void parity_destroy(void* parity);
 
-/// Performs an RPC request.
+/// Performs an asynchronous RPC request running in a background thread for at most X milliseconds
+///
+///	- parity		: Reference to the running parity client
+///	- rpc_query		: JSON encoded string representing the RPC request.
+///	- len			: Length of the RPC query
+///	- timeout_ms	: Maximum time that request is waiting for a response
+///	- response		: Callback to invoke when the query gets answered. It will respond with
+///
+///						1) A JSON encoded string with the result
+///						2) A string "empty", (got an empty response)
+//						3) A string "timeout", (the query timed-out)
 ///
-/// Blocks the current thread until the request is finished. You are therefore encouraged to spawn
-/// a new thread for each RPC request that requires accessing the blockchain.
+/// - On success	: The parity client reference and the query string were valid
+/// - On error		: The parity client reference and the query string were not valid
 ///
-/// - `rpc` and `len` must contain the JSON string representing the RPC request.
-/// - `out_str` and `out_len` point to a buffer where the output JSON result will be stored. If the
-///	  buffer is not large enough, the function fails.
-/// - `out_len` will receive the final length of the string.
-/// - On success, the function returns 0. On failure, it returns 1.
+int parity_rpc(const void *const parity, const char* rpc_query, size_t rpc_len, size_t timeout_ms, callback response);
```

Comment:
Are we talking about distinguishing between `ordinary RPC queries` and `websocket subscriptions in this context? If so, then I would expect people to use to different callbacks but sure I can add it anyway probably

When it comes to actual WebSocket responses I just use `a dummy regex` to get the subscription id which must be used to disable the subscription! I expect people to use some JSON parser in that case

Reply:
> to use to different callbacks but sure I can add it anyway probably

Well, you can't really do that correctly in C. The "callback-with-a-void*-parameter" pattern is very common in C.

Reply:
hmm, I don't understand your comment, you should be able this this in C right?

```c
static void f1(void) {
    printf("f1\r\n");
}

static void f2(void) {
    printf("f2\r\n");
}

void register_callback(void (*callback)(void))
{
    printf("inside register_callback");
    (callback)();
}

int main(void) {
    register_callback(f1);
    register_callback(f2);
}
```

Or what do you mean? :P

Reply:
Imagine you want to query the block header of a certain block number.
In the callback, you parse the JSON response and you obtain the header and number. But then you probably want to do with that information, but since you don't have any context you have no other way than to use global variables. This is not great.

Instead, if you have this `void*` parameter you can pass any information required for the callback to do what it needs to do.